### PR TITLE
fix DexWriter for hiddenapi section

### PR DESCRIPTION
**Pull request by robertsmd**
solves #815

### CODE REVIEW

File: dexlib2/src/main/java/org/jf/dexlib2/writer/DexWriter.java [hunk 1]
```diff
@@ -221,6 +221,16 @@ public int compare(Entry<? extends CallSiteKey, Integer> o1, Entry<? extends Cal
...
+    private static <T extends Comparable<? super T>> Comparator<Map.Entry<? extends T, ?>> comparableValueComparator() {
+        return new Comparator<Entry<? extends T, ?>>() {
+            @Override public int compare(Entry<? extends T, ?> o1, Entry<? extends T, ?> o2) {
+                int o1_int = Integer.parseInt(o1.getValue().toString());
+                int o2_int = Integer.parseInt(o2.getValue().toString());
+                return Integer.compare(o1_int, o2_int);
+            }
+        };
+    }
+
```
Comment (Lanchon) [inline dexlib2/src/main/java/org/jf/dexlib2/writer/DexWriter.java:233]: why the trip to string? are the values not all Integer? then create a `comparableIntegerValueComparator` and type it accordingly.

better yet, make `comparableValueComparator` really generic. what is wrong with this?
```
<T extends Comparable<? super T>> Comparator<Map.Entry<?, ? extends T>> comparableValueComparator() {
        return new Comparator<Entry<?, ? extends T>>() {
            @Override public int compare(Entry<?, ? extends T> o1, Entry<?, ? extends T> o2) {
                return o1.getValue().compareTo(o2.getValue());
            }
        }
    }
```

File: dexlib2/src/main/java/org/jf/dexlib2/writer/DexWriter.java [hunk 2]
```diff
@@ -503,8 +513,11 @@ private void writeClasses(@Nonnull DexDataStore dataStore, @Nonnull DexDataWrite
             return;
         }
 
+        offsetWriter.align();
         hiddenApiRestrictionsOffset = offsetWriter.getPosition();
 
+        Collections.sort(classEntries, DexWriter.<ClassKey>comparableValueComparator());
+
```
Comment (Lanchon) [inline dexlib2/src/main/java/org/jf/dexlib2/writer/DexWriter.java:520]: what are the consequences of leaving `classEntries` with an order that depends on `shouldWriteHiddenApiRestrictions()`? could sorting a copy be preferable?

Comment (robertsmd) [reply dexlib2/src/main/java/org/jf/dexlib2/writer/DexWriter.java:520]: yeah that would be fine, I believe. it's just for the hiddenapi section.

File: dexlib2/src/main/java/org/jf/dexlib2/writer/DexWriter.java (hunk 1 above)
Comment (robertsmd) [reply dexlib2/src/main/java/org/jf/dexlib2/writer/DexWriter.java:233]: Agreed. I like this better.

**Comment by robertsmd** [comment]
successfully pushed refactors based on @Lanchon 's suggestions. Apologies for the bad casts and whatnot, java isn't my main language so I was attempting the functionality to work on a first pass and didn't cleanup.

**Comment by koxu1996** [comment]
@JesusFreke Will it be merged?

**Comment by SolidHal** [comment]
+1 to getting this merged, confirmed this fixes the issue

**Comment by dukejob** [comment]
@JesusFreke This pr is really helpful

**Comment by thanhpv15** [comment]
Please review this PR, it's important for Apktool's issue

**Comment by JesusFreke** [comment]
Thanks. Squashed and merged as 3436508beafbbacf3aae5543cdd1f1eb5edf5017

**Pull request closed without merging**